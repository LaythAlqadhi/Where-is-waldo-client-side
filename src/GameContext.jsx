import React, { useState, createContext } from 'react';
import PropTypes from 'prop-types';

const initialState = {
  state: { loading: false, error: null },
  status: { isStarted: false, isEnded: true },
  message: null,
};

export const GameContext = createContext(initialState);

export function GameProvider({ children }) {
  const [state, setState] = useState({ loading: false, error: null });
  const [status, setStatus] = useState({
    isStarted: false,
    isEnded: false,
    message: null,
  });
  const [message, setMessage] = useState(null);

  return (
    <GameContext.Provider
      value={{
        state,
        setState,
        status,
        setStatus,
        message,
        setMessage,
      }}
    >
      {children}
    </GameContext.Provider>
  );
}

GameProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
