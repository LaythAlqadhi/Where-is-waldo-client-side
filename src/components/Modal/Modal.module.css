.modal {
  max-width: 600px;
  width: 80%;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  background-color: var(--primary-color);
  padding: 50px;
  border-radius: 16px;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.5);
}

.modal form {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.modal form label {
  align-self: flex-start;
  font-size: 1.25rem;
  margin-bottom: 16px;
}

.modal form input {
  margin-bottom: 16px;
  padding: 16px 8px;
  width: 100%;
  font-size: 1.25rem;
  border: 2px solid var(--secondary-color);
  border-radius: 4px;
  background-color: var(--primary-color);
  color: var(--secondary-color);
}

.modal form span {
  font-size: 1.25rem;
  color: red;
}

.modal form div {
  display: flex;
}