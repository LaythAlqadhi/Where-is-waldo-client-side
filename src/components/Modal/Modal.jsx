import React, { useState, useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import PropTypes from 'prop-types';
import { GameContext } from '../../GameContext';
import style from './Modal.module.css';

function Modal({ id }) {
  const navigate = useNavigate();
  const { state, setState, status, setStatus } = useContext(GameContext);
  const [name, setName] = useState('');
  const [message, setMessage] = useState(null);

  const handleClick = (e, submit) => {
    e.preventDefault();

    if (submit && name === '') {
      setMessage('Name must be not empty.');
      return;
    }
    if (name.length > 25) {
      setMessage('Name must be less than 25 characters.');
      return;
    }

    const fetchData = async () => {
      const API_URL = 'https://node.laythalqadhi.repl.co/v1';

      try {
        setState({ ...state, loading: true });

        const response = await fetch(`${API_URL}/end/${id}`, {
          method: 'POST',
          mode: 'cors',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            name: name || null,
          }),
        });

        if (response.status >= 400) {
          throw new Error('Server error');
        }
      } catch (err) {
        setState({ ...state, error: err.message });
      } finally {
        setName('');
        setMessage(null);
        setState({ ...state, loading: false });
        setStatus({ ...status, isStarted: false, isEnded: false });

        if (submit) {
          navigate('/leaderboard');
        } else {
          navigate('/');
        }
      }
    };
    fetchData();
  };

  return (
    <div className={style.modal}>
      <form>
        <label htmlFor="user">
          Please enter your name to be included in the leaderboard.
        </label>
        <input
          type="text"
          name="user"
          id="user"
          placeholder="John Doe"
          maxLength="25"
          value={name}
          onChange={(e) => setName(e.target.value)}
        />
        <div>
          <button type="button" onClick={(e) => handleClick(e, true)}>
            Submit
          </button>
          <button type="button" onClick={(e) => handleClick(e, false)}>
            Cancel
          </button>
        </div>
        {message && <span>{message}</span>}
      </form>
    </div>
  );
}

Modal.propTypes = {
  id: PropTypes.string.isRequired,
};

export default Modal;
