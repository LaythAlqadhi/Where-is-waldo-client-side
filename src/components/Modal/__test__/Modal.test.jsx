import React from 'react';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';
import { GameContext } from '../../../GameContext';
import Modal from '../Modal';

it('renders Modal component correctly', async () => {
  render(
    <MemoryRouter>
      <GameContext.Provider
        value={{
          state: { loading: false, error: null },
          status: { isStarted: true, isEnded: false },
          setState: vi.fn(),
          setStatus: vi.fn(),
        }}
      >
        <Modal id="mockId" />
      </GameContext.Provider>
    </MemoryRouter>,
  );

  expect(screen.container).toMatchSnapshot();

  userEvent.type(screen.getByPlaceholderText('John Doe'), 'John Doe');

  expect(screen.container).toMatchSnapshot();

  userEvent.click(screen.getByText('Submit'));

  expect(screen.container).toMatchSnapshot();

  userEvent.click(screen.getByText('Cancel'));

  expect(screen.container).toMatchSnapshot();
});
