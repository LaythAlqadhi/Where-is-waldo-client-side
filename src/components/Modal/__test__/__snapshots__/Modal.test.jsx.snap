// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal component renders correctly 1`] = `undefined`;

exports[`Modal component renders correctly 2`] = `undefined`;

exports[`Modal component renders correctly 3`] = `undefined`;

exports[`Modal component renders correctly 4`] = `undefined`;

exports[`renders Modal component correctly 1`] = `undefined`;

exports[`renders Modal component correctly 2`] = `undefined`;

exports[`renders Modal component correctly 3`] = `undefined`;

exports[`renders Modal component correctly 4`] = `undefined`;
